src/include/migraphx/shape_for_each.hpp

```diff
@@ -51,9 +51,10 @@ void shape_for_each(const migraphx::shape& s, F f)
                            assert(len > 0 and stride > 0);
                            return (i / stride) % len;
                        });
-        call(indices);
+        f(i, index_const_ref);
```

Comment:
The `i` should be optional since its only used in a few places, and should be the last parameter. You can create an invoker to call the function like this:

```cpp
template<class F, class Vector>
auto shape_for_each_invoke(rank<2>, F& f, const Vector& v, std::size_t i) -> decltype(f(v, i))
{
    return f(v, i);
}

template<class F, class Vector>
auto shape_for_each_invoke(rank<1>, F& f, const Vector& v, std::size_t) -> decltype(f(v))
{
    return f(v);
}
``` 

And then you can do `shape_for_each_invoke(rank<2>{}, f, index_const_ref, i)` to call the function.

Reply:
I first implemented it by making it the second parameter, just because it is optional in a few instances. But then felt it was logical to make it the first one: <an index, followed by its vector representation>. Hence this way.. 

(Personally, I am not a fan of putting a rank<int>, just to invoke a simple function).

Reply:
BTW, if a template were to be used, we could do it without a rank..;

Reply:
BTW, `get_original_idx_op()` in parse_resize.cpp does exactly this: its first few parameters in the lambdas are just ignored.. same style here.. Thanks.

Reply:
> Personally, I am not a fan of putting a rank, just to invoke a simple function

Rank is to avoid ambiguous overloads. Most likely it wont be needed, but if the we were to pass a variadic lambda in the future it would fail. Perhaps the error is acceptable as a variadic lambda does seem a little ambiguous in this context.

So I think the `rank` can be removed from the functions.

Reply:
Thanks, Paul. We have a very simple case here where it would be fairly routine stuff for a decent compiler to warn if something was amiss -- e.g. unused parameter. I think we can worry about the future lambda cases later. 

My recommendation is to leave this change as it is, until somebody changes something in future about the variadic lambda(s).

Reply:
We should still make the parameter optional(even though we dont use `rank` here), just so this doesn't need to be changed everywhere `shape_for_each` is used and it wont break other developers who may be using `shape_for_each` in their unmerged work or PRs.

Reply:
Without `rank`:

```cpp
template<class F, class Vector>
auto shape_for_each_invoke(F& f, const Vector& v, std::size_t i) -> decltype(f(v, i))
{
    return f(v, i);
}
template<class F, class Vector>
auto shape_for_each_invoke(F& f, const Vector& v, std::size_t) -> decltype(f(v))
{
    return f(v);
}
```

Reply:
Would this be okay? I compiled this -- in shape_for_each.hpp: 
```
template <class F, class Vec = std::vector<std::size_t>, class Idx = size_t>
void shape_for_each(const migraphx::shape& s, F f)
{
    std::vector<std::size_t> indices(s.lens().size());
    auto const& index_const_ref = indices;
    shape ss{s.type(), s.lens()};
    size_t max = ss.elements();
    for(std::size_t i = 0; i < max; i++)
    {
        std::transform(ss.strides().begin(),
                       ss.strides().end(),
                       ss.lens().begin(),
                       indices.begin(),
                       [&](std::size_t stride, std::size_t len) {
                           assert(len > 0 and stride > 0);
                           return (i / stride) % len;
                       });
        if constexpr(std::is_invocable_v<F, Vec, Idx>)
            f(index_const_ref, i);
        else
            f(index_const_ref);
    }
}
```

Reply:
(Above) Code is checked in. But why have some build process failed -- particularly the 20.04 debug build -- it isn't clear to me.

Reply:
Yes that would work, but I would prefer it to be written as `std::is_invocable<F, decltype(index_const_ref), decltype(i)>{}`.

Reply:
> But why have some build process failed -- particularly the 20.04 debug build -- it isn't clear to me.

Its a network issue, I dont think its an issue with your code.